src/sentry/rules/processing/buffer_processing.py

```diff
@@ -0,0 +1,155 @@
+class BufferHashKeys:
+    model: type[models.Model]
+    filters: FilterKeys
+
+
+class DelayedProcessingBase(ABC):
+    buffer_key: ClassVar[str]
+
+    def __init__(self, project_id: int):
+        self.project_id = project_id
+
+    @property
+    @abstractmethod
+    def hash_args(self) -> BufferHashKeys:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def processing_task(self) -> Task:
+        raise NotImplementedError
+
+
+delayed_processing_registry = Registry[type[DelayedProcessingBase]]()
+
+
+def fetch_alertgroup_to_event_data(
+    project_id: int, model: type[models.Model], batch_key: str | None = None
+) -> dict[str, str]:
+    field: dict[str, models.Model | int | str] = {
+        "project_id": project_id,
+    }
+
+    if batch_key:
+        field["batch_key"] = batch_key
+
+    return buffer.backend.get_hash(model=model, field=field)
+
+
+def bucket_num_groups(num_groups: int) -> str:
+    if num_groups > 1:
+        magnitude = 10 ** int(math.log10(num_groups))
+        return f">{magnitude}"
+    return "1"
+
+
+def process_in_batches(project_id: int, processing_type: str) -> None:
+    """
+    This will check the number of alertgroup_to_event_data items in the Redis buffer for a project.
+
+    If the number is larger than the batch size, it will chunk the items and process them in batches.
+
+    The batches are replicated into a new redis hash with a unique filter (a uuid) to identify the batch.
+    We need to use a UUID because these batches can be created in multiple processes and we need to ensure
+    uniqueness across all of them for the centralized redis buffer. The batches are stored in redis because
+    we shouldn't pass objects that need to be pickled and 10k items could be problematic in the celery tasks
+    as arguments could be problematic. Finally, we can't use a pagination system on the data because
+    redis doesn't maintain the sort order of the hash keys.
+
+    `processing_task` will fetch the batch from redis and process the rules.
+    """
+    batch_size = options.get("delayed_processing.batch_size")
+    log_format = "{}.{}"
+
+    try:
+        processing_info = delayed_processing_registry.get(processing_type)(project_id)
+    except NoRegistrationExistsError:
+        logger.exception(log_format.format(processing_type, "no_registration"))
+        return
+
+    hash_args = processing_info.hash_args
+    task = processing_info.processing_task
+    filters: dict[str, models.Model | str | int] = asdict(hash_args.filters)
+
+    event_count = buffer.backend.get_hash_length(model=hash_args.model, field=filters)
+    metrics.incr(
+        f"{processing_type}.num_groups", tags={"num_groups": bucket_num_groups(event_count)}
+    )
+
+    if event_count < batch_size:
+        return task.delay(project_id)
+
+    logger.info(
+        log_format.format(processing_type, "process_large_batch"),
+        extra={"project_id": project_id, "count": event_count},
+    )
```

Comment:
should we gate this log with a flag or anything? that way we don't send a log per batch evaluation?

Reply:
done